Implement repo level, user level and global environment variables support

dvjn opened the pull request:
Fixes #3661

dvjn commented:
Hi @6543 @anbraten @qwerty287 I have done initial UI changes for this, please review the proposal and screenshots and let me know if I should continue working on this and implement the rest.


**Screenshots:**

Repo Settings:
![repo settings](https://github.com/woodpecker-ci/woodpecker/assets/62170586/c2ca7d48-acec-4344-b790-49fb44062432)

User Settings:
![user settings](https://github.com/woodpecker-ci/woodpecker/assets/62170586/0d0930be-8e7f-4410-9e73-48f1293a32b6)

Admin Settings:
![admin settings](https://github.com/woodpecker-ci/woodpecker/assets/62170586/7aae116d-119e-46ad-a058-f95731ed3361)

Add Variable Screen:
![add variable](https://github.com/woodpecker-ci/woodpecker/assets/62170586/cacbe4cd-f92a-4fcb-8287-8972ea97df60)

anbraten commented:
As variables are not secret anyways there was a suggestion somewhere to have a special config file like `.woodpecker/config.yaml` in the repo which can be used for such settings. This would be quite handy as it would also allow to use those variables in the `cli exec` command as well and allows easy copies of that config to other woodpecker instances etc. Should those variables then be shown in the UI as well as read-only or should it only be the config file? What do you think?

zc-devs commented:
JFYI, how it is implemented in [Gitea](https://blog.gitea.com/release-of-1.22.0/#rocket-support-instance-level-actions-variables-28115).

dvjn commented:
> As variables are not secret anyways there was a suggestion somewhere to have a special config file like `.woodpecker/config.yaml` in the repo which can be used for such settings.

I think this is not as useful as it seems, since you can already use variables in a workflow file. This `config.yaml` approach is mainly beneficial when you need these variables in multiple workflows within the same repository.
However, the most important use-cases of this PR, in my opinion, are: first, the ability to manage global variables using the admin UI, and second, the capability to share variables across repositories under an organization or user.

FYI, the solution I'm proposing is heavily inspired by Gitea.

dvjn commented:
@anbraten @qwerty287 @zc-devs
Do let me know your final thoughts on this and let me know if I should move forward with implementing it.

dvjn commented:
I can also add agent variables to address #3311

zc-devs commented:
> your final thoughts

I like the idea.

> if I should move forward with implementing it

I am not a right guy :) Let's hear from maintainers.

> I can also add agent variables

That would be great!

dvjn commented:
> However, you write "Fixes #3661" in the desc which is not correct (I believe there's an issue about your thing too). #3661 is about workflow-level vars and this should definitely be defined in the yaml.

My bad, will remove that issue from the description.

itsTurnip commented:
I'm really interested in this PR. Keep moving! Thanks for your work.

dvjn commented:
I have completed the code changes in the server and web frontend. The docs and cli changes are still pending. Planning to do those after an initial round of code review.

It would be great if someone could review the code.

Summary of Changes:
- Duplicated code for secrets (removing events and images).
- Added support for `from_variable` and `variables` (similar to `from_secret` and `secrets`) in pipeline config.
- Added new table in database `variables`, with migrations.
- Copied some secrets test cases and used for variables.

lafriks commented:
Wasn't there an idea to drop `secrets` section from step? Maybe let's do not add variables section and keep only one way to use them with `from_variable`?

dvjn commented:
> Wasn't there an idea to drop `secrets` section from step?

I am not aware of it. If it is just an idea without a proper discussion, maybe let's not act on that?

> Maybe let's do not add variables section and keep only one way to use them with `from_variable`?

I have kept feature parity with secrets, to not cause confusion. If we want to remove it, wouldn't it be better to remove both `secrets` and `variables` together in a separate PR.

Just my thoughts, let me know if we still want to remove it.

zc-devs commented:
If I set some variable
![333869343-cacbe4cd-f92a-4fcb-8287-8972ea97df60](https://github.com/user-attachments/assets/fda2b1ad-b516-4864-b9d1-14054942b9e4)
and run pipeline
```yaml
skip_clone: true
steps:
  glob-vars:
    image: alpine
    commands:
      - echo $$COMPANY
```
what would I see in steps (UI) logs?

--- 

Maybe add `build_pr_images` label?

dvjn commented:
> what would I see in steps (UI) logs?

It would be empty as you haven't explicitly added those variables in your step either by using `variables` or `from_variable`.

zc-devs commented:
As this is not a secret value, it doesn't make sense to restrict this values by a step (to me).
So, I would expect it to be printed without `explicitly added those variables in your step`.
In case if I want to map "global" var to different name in a step, I can use `from_variable`.

dvjn commented:
@zc-devs

That is a good point. That would make this feature much easier to use.

Although, I feel that this kind of implicit behaviour might cause some confusion. It won't always be obvious where an environment variable is coming from.

Also, it could be confusing for a user who has push permissions on a repo, but cannot change the repo settings. This kind of user won't be able to view where the variables are automatically getting set from.

dvjn commented:
I understand your point, but I'm still afraid of adding this new implicit behavior.

Can we get another opinion if that's okay?

dvjn commented:
Hi maintainers, it would be great to get a review from y'all.

lafriks commented:
I don't think setting them explicitly is a good idea as that could have unforeseen consequences that would be not so easy to track down both from usability and security perspective

lafriks commented:
> > Wasn't there an idea to drop `secrets` section from step?
> 
> I am not aware of it. If it is just an idea without a proper discussion, maybe let's not act on that?
> 
> > Maybe let's do not add variables section and keep only one way to use them with `from_variable`?
> 
> I have kept feature parity with secrets, to not cause confusion. If we want to remove it, wouldn't it be better to remove both `secrets` and `variables` together in a separate PR.
> 
> Just my thoughts, let me know if we still want to remove it.

Personally I don't see benefit of having multiple ways of using same feature, just adds confusion.

Also if we add it and later on remove we will have to again create more breaking changes that can be prevented without adding that in the first place

lafriks commented:
I don't think setting them explicitly is a good idea as that could have unforeseen consequences that would be not so easy to track down both from usability and security perspective

lafriks commented:
Also would be confusing if both variables and secrets would be used by the same name on what takes preference etc.

I only briefly skimmed through file changes but imho this introduces back security vulnerability with using `variables` section with plugins

dvjn commented:
Thank you for the comments.

So from what I understood, I'll just keep one way to use variables, i.e., using `from_variable`, and remove support for `variables` in pipeline. Right?

Then I'll complete the Cli and Docs changes for the final review.

dvjn commented:
@lafriks I have removed the support for using `variables`, and completed the go client, cli and doc changes.

This PR is ready to be reviewed and merged.

dvjn commented:
Hi maintainers, I have finished working on this PR. This is ready for review.

6543 commented:
before reading more into this pull code and issue, tow important questions:
- why did you touch the `pipeline/*` code at all? I thought we already allow to have environment viariables
- if variables dont have a prefix, all currently steps counted as plugins wo have variables set should not be contued as plugins!

6543 commented on pipeline/frontend/yaml/compiler/convert.go:
```diff
@@ -98,6 +98,16 @@ func (c *Compiler) createProcess(container *yaml_types.Container, stepType backe
 	workingDir = c.stepWorkingDir(container)
 
+	getVariableValue := func(name string) (string, error) {
+		name = strings.ToLower(name)
```
we switched back to case sensitive for secrets ... so we should stay consistent

dvjn replied:
I can still see `name = strings.ToLower(name)` in the `getVariableValue` even after merging main, should I still proceed to remove them from variables?

6543 replied:
let me have a look at the last pulls in this topic ...

dvjn replied:
Any updates on this?

6543 replied:
yes case sensitive it is

6543 commented on web/src/assets/locales/en.json:
```diff
@@ -255,6 +255,9 @@
         "desc": "Organization secrets can be passed to all organization's repository's pipeline steps as environmental variables."
       },
+      "variables": {
+        "desc": "Organization variables can be passed to all organization's repository's pipeline steps as environmental variables."
```
```suggestion
        "desc": "Organization variables can be passed to all organization's repository's pipeline steps (e.g. as environmental variables)."
```

6543 commented on web/src/assets/locales/en.json:
```diff
@@ -409,6 +419,22 @@
   },
+  "variables": {
+    "variables": "Variables",
+    "desc": "Variables can be passed to individual pipeline steps at runtime as environmental variables.",
```
```suggestion
    "desc": "Variables can be passed to individual pipeline steps at execution start.",
```

6543 replied:
clear differentiate to variables as e.g. github-actions uses them

6543 commented:
uh damed i got requested but it fall under my todo list :?

6543 commented on pipeline/frontend/yaml/compiler/compiler.go:
```diff
@@ -107,6 +113,7 @@ func New(opts ...Option) *Compiler {
 		env:                 map[string]string{},
 		cloneEnv:            map[string]string{},
 		secrets:             map[string]Secret{},
+		variables:           map[string]Variable{},
```
I think a simple `map[string]string` should be sufficient enouth ...

6543 replied:
bf7b2996c4c530a4e2d8cc95d12f37f4eec56caf

6543 commented on pipeline/frontend/yaml/compiler/settings/params.go:
```diff
@@ -160,10 +160,10 @@ func handleComplex(v any, getSecretValue func(name string) (string, error)) (str
 // else it just indicates to progress normally using the provided map as is.
-func injectSecret(v map[string]any, getSecretValue func(name string) (string, error)) (string, bool, error) {
+func injectVariableOrSecret(v map[string]any, getVariableValue, getSecretValue func(name string) (string, error)) (string, bool, error) {
 	if secretNameI, ok := v["from_secret"]; ok {
```
we might move those two keys into const

6543 replied:
4c4221e80f0f4651bb00f69b0a4f8fa9932c3701

6543 commented:
Review process:
 - [x] docs
 - [x] pipeline engine
 - [ ] api
 - [ ] server datastore
 - [ ] cli
 - [ ] webui

6543 commented:
NOTE: write into docs that it is case-sensitive

6543 commented:
# TODO:
 - [ ] The Variable views should addopt KeyValueEditor (https://github.com/woodpecker-ci/woodpecker/pull/4412)
 - [ ] variable creation via WebUI should work again

6543 commented:
@dvjn using secrets as template is fine ... but there are a lot of copy-paste errors :/

pat-s commented:
> It would be empty as you haven't explicitly added those variables in your step either by using variables or from_variable.

If that would be the case, most of the implementation would be meaningless as then one could simply use `secrets`, as already mentioned.

The real value of this PR IMO lies in the possibility to (easily) set global/org-wide variables which are not sensitive. Without having to repeat them within the YAML again.
For every variable there could/should be an informative log line telling that this variable has been set by a global/org-wide one.

Also, I am not sure if "variables" is the best wording if the values are mapped as env vars in the end. Then, it could directly be named "Environment". This might avoid confusion if there is a difference between a "variable" and an "env var".
